### PR TITLE
Refactor and improve archiving/publishing rules

### PR DESCRIPTION
**Pull request by mickaelandrieu**
* Move validation to his own methods for publishing and archiving
* Apply some exceptions on business rules
* Added unit tests

### CODE REVIEW

**Comment by mickaelandrieu** [comment]
@eric-chau please review :)

File: NestedNode/Page.php [hunk 1]
```diff
@@ -952,12 +954,31 @@ public function setState($state)
...
+     */
+    private function validatePublishing($publishing)
+    {
         if ($this->isRoot() && $publishing !== null && $this->_publishing !== null) {
             throw new \LogicException("Root page is already published.");
         }
-        $this->_publishing = null !== $publishing ? $this->convertTimestampToDateTime($publishing) : null;
 
-        return $this;
+        if ($publishing < new \DateTime()) {
```
Comment (eric-chau) [inline NestedNode/Page.php]: Can we be more precise like something like this?
```php
if ($publishing instanceof \DateTime && $publishing->getTimestamp() < time()) 
```

**Comment by eric-chau** [comment]
Nice contribution; thanks :+1:

**Comment by mickaelandrieu** [comment]
Thanks for review :) if tests pass, I'll merge.

**Pull request merged**